### ISSUE TITLE
Create new unit tests for Evennia - hacktoberfest

### ISSUE THREAD

**Issue by Griatch** [evennia/evennia#1458]
#### Brief summary of issue / Description of requested feature:

This is a [hacktoberfest](https://hacktoberfest.digitalocean.com/) issue for expanding Evennia's unit test coverage. 

#### Steps to reproduce the issue / Reasons for adding feature:

1. Go to [the coveralls page for Evennia](https://coveralls.io/github/evennia/evennia) 
2. Find an Evennia module under `evennia/` that lacks unit test coverage 
3. If no `tests.py`file exists in the sub-package in question, add it. See other `tests.py` in the Evennia repo for help.

#### Expected result of feature

We are currently at 53% unit test coverage, why not expand that? 

#### Extra information, such as Evennia revision/repo/branch, operating system and ideas for how to solve / implement:

This is against `master` branch.

**Comment by vaibhavsingh97** [comment]
i can give it a try 😄

**Comment by Keda87** [comment]
Hi, just find out this repo for hacktoberfest.
I'll try to add some test tonight, do you have any convention/documentation to create test and PR?

**Comment by Griatch** [comment]
@Keda87 That's great! I updated the issue with some more info. :)

**Comment by Griatch** [comment]
I also added some exceptions. No need to dig into making unit tests for the game_template since it's basically well, empty template files.

**Comment by Keda87** [comment]
@Griatch how can I see coverage report for testing on my local machine?

**Comment by Griatch** [comment]
@keda87 It's not necessary to run `coverage` locally for this issue. But if you want to try it you need to follow these steps (only tested on Linux): 

1. In your virtualenv, do  `pip install coverage`
2. Assuming you followed Evennia according to the `Getting Started` instructions you should have your `evennia/` folder sitting next to your game dir (called e.g. `mygame/`). `cd` into your game dir.
3. Run one of the following:        
```bash
# linux 
coverage run --source=../evennia --omit=*/migrations/*,*/urls.py,*/test*.py,*.sh,*.txt,*.md,*.pyc,*.service ../evennia/bin/unix/evennia test evennia

# windows (untested)
coverage run --source=../evennia --omit=*/migrations/*,*/urls.py,*/test*.py,*.sh,*.txt,*.md,*.pyc,*.service ../evennia/bin/windows/evennia test evennia
```
4. Let the tests run to completion. Then run `coverage report`. You will get a list of all files and the full unit test coverage of the entire Evennia package (currently 53%). You can also do `coverage report path/to/python/file.py` to see coverage of only that file.

**Comment by gulasnani** [comment]
@Griatch Hi, I read some documentation and have started with some testing on syscommands.py, but I am not able to follow how the args passed on to cmdobj in each test are used. Could you direct me to the particular docs that may make me understand how the strings passed in the tests as args handled by CommandTest?

Spare me for a noobish question but am new to gitHub :p 
Thanks!

**Comment by Keda87** [comment]
@Griatch thank you, I just want to ensure the coverage increasing

**Comment by Griatch** [comment]
@gulasnani As you can see in Testcmd, the `args` you pass in is stored on the command object here https://github.com/evennia/evennia/blob/master/evennia/commands/default/tests.py#L64. 

Normally (when a real command is run in Evennia), these are the arguments passed to the command. So if the command was `look here` (defined by the CmdLook class in another module) then `args` would be the string `" here"`.  

Now, in the case of the test, the execution of the command is emulated, with the body of the command being executed when first the `.parse` and then the `.func` method fires here: https://github.com/evennia/evennia/blob/master/evennia/commands/default/tests.py#L76 (in a live Evennia, this sequence is called by the cmdhandler system). As `.parse` and then  `.func` runs, the developer creating a Command has access to all properties assigned to the Command instance earlier, including that `.args`. How `.parse` and `.func` are implemented is up to the developer and what kind of command is desired. In the example of `look here`, it would try to figure out what `"here"` means and try to display its description. 

In the specific case of the `syscommands.py` module, those commands are standard examples of various fallback commands to run when no specific one has been found. So they often don't do much by default. But here is an example of one of them picking up the `.args` and operating on it: https://github.com/evennia/evennia/blob/master/evennia/commands/default/syscommands.py#L149

Hope that helps!

**Comment by gulasnani** [comment]
@Griatch Thanks a lot for walking me through! Quite helpful it was, as there were just a few bits I overlooked which were of actual importance for the overall understanding.

**Comment by vonzimr** [comment]
Is there a nice way to run only specific tests? I see that I can run
``evennia test evennia``
or 
``evennia test [component]``

But if I just wanted to test, say, the accounts module how would I go about that?

**Comment by vonzimr** [comment]
Reading the django docs, I found writing
``evennia test evennia.[module name]``
seems to work!

**Comment by Griatch** [comment]
Yes `evennia test enennia.accounts` will work. To get to a specific test in a tests.py module you could also specify down to class or even class method level; like  `evennia test evennia.accounts.tests.TestStuff.test_particular_thing`.

**Comment by BeaData** [comment]
Hi, I'd like to help too by running some tests :)
Currently setting up Evennia; I'll work on Windows 10 by the way, any particular issues you may want me to focus on? Let me know!

**Comment by Griatch** [comment]
@BeaData Any tests are good tests. See the first post of this issue, they should help to get you going. Pick an untested module/package that looks interesting to you. :)

**Comment by selberhad** [comment]
@Griatch There seems to be an issue with the coveralls configuration. I can't view any of the source files to see where the coverage gaps are. The error is of the form:

```
SOURCE NOT AVAILABLE
The file "/home/runner/work/evennia/evennia/evennia/contrib/tutorial_world/objects.py" isn't available on github. Either it's been removed, or the repo root directory needs to be updated.
```
Unless I'm doing something, it's been a few years since I last used coveralls.

**Comment by Griatch** [comment]
@selberhad Oh, haven't looked at the detailed coverage in the while. I updated the root now, so you should be able to see the source code. Thanks for the heads up! 

(The `develop` branch coverage on coveralls.io is also weird. If you run locally, the coverage for that branch  is 66%, but on coveralls.io it reports 26% for develop branch. It just dropped some months ago. Not sure what's going on there :frowning_face: )

**Comment by Griatch** [comment]
Turns out that coverage doesn't like parallel execution. The coverage data was fixed now, so `develop` branch shows the correct percentage (66%): https://coveralls.io/github/evennia/evennia?branch=develop :tada:
